File: src/main/java/me/ag2s/epublib/domain/Identifier.java
package me.ag2s.epublib.domain;

import java.io.Serializable;
import java.util.List;
import java.util.UUID;
import me.ag2s.epublib.util.StringUtil;

public class Identifier implements Serializable {
   private static final long serialVersionUID = 955949951416391810L;
   private boolean bookId;
   private String scheme;
   private String value;

   public Identifier() {
      this("UUID", UUID.randomUUID().toString());
   }

   public Identifier(String scheme, String value) {
      this.bookId = false;
      this.scheme = scheme;
      this.value = value;
   }

   public static Identifier getBookIdIdentifier(List<Identifier> identifiers) {
      if (identifiers != null && !identifiers.isEmpty()) {
         Identifier result = null;

         for(Identifier identifier : identifiers) {
            if (identifier.isBookId()) {
               result = identifier;
               break;
            }
         }

         if (result == null) {
            result = (Identifier)identifiers.get(0);
         }

         return result;
      } else {
         return null;
      }
   }

   public String getScheme() {
      return this.scheme;
   }

   public void setScheme(String scheme) {
      this.scheme = scheme;
   }

   public String getValue() {
      return this.value;
   }

   public void setValue(String value) {
      this.value = value;
   }

   public void setBookId(boolean bookId) {
      this.bookId = bookId;
   }

   public boolean isBookId() {
      return this.bookId;
   }

   public int hashCode() {
      return StringUtil.defaultIfNull(this.scheme).hashCode() ^ StringUtil.defaultIfNull(this.value).hashCode();
   }

   public boolean equals(Object otherIdentifier) {
      if (!(otherIdentifier instanceof Identifier)) {
         return false;
      } else {
         return StringUtil.equals(this.scheme, ((Identifier)otherIdentifier).scheme) && StringUtil.equals(this.value, ((Identifier)otherIdentifier).value);
      }
   }

   public String toString() {
      return StringUtil.isBlank(this.scheme) ? "" + this.value : "" + this.scheme + ":" + this.value;
   }

   public interface Scheme {
      String UUID = "UUID";
      String ISBN = "ISBN";
      String URL = "URL";
      String URI = "URI";
   }
}

File: src/main/kotlin/me/ag2s/epublib/domain/Identifier.kt
package me.ag2s.epublib.domain

import java.io.Serializable
import java.util.UUID
import me.ag2s.epublib.util.StringUtil

class Identifier : Serializable {
    private var bookId = false
    private var scheme: String?
    private var value: String?
    constructor() : this("UUID", UUID.randomUUID().toString())
    constructor(scheme: String?, value: String?) { this.scheme = scheme; this.value = value }
    fun getScheme(): String? = scheme
    fun setScheme(scheme: String?) { this.scheme = scheme }
    fun getValue(): String? = value
    fun setValue(value: String?) { this.value = value }
    fun setBookId(bookId: Boolean) { this.bookId = bookId }
    fun isBookId(): Boolean = bookId
    override fun hashCode(): Int = StringUtil.defaultIfNull(scheme).hashCode() xor StringUtil.defaultIfNull(value).hashCode()
    override fun equals(other: Any?): Boolean = other is Identifier && StringUtil.equals(scheme, other.scheme) && StringUtil.equals(value, other.value)
    override fun toString(): String = if (StringUtil.isBlank(scheme)) "" + value else "" + scheme + ":" + value
    interface Scheme { companion object { const val UUID = "UUID"; const val ISBN = "ISBN"; const val URL = "URL"; const val URI = "URI" } }
    companion object { private const val serialVersionUID = 955949951416391810L; @JvmStatic fun getBookIdIdentifier(identifiers: List<Identifier>?): Identifier? { if (identifiers.isNullOrEmpty()) return null; return identifiers.firstOrNull { it.isBookId() } ?: identifiers[0] } }
}
